Panic on close of closed channel

The following happens when stopping the all-in-one when badger is being used as storage, probably related to #2076:

```console
$ SPAN_STORAGE_TYPE=badger go run -tags ui ./cmd/all-in-one 
...
^C{"level":"info","ts":1582817948.1366558,"caller":"flags/service.go:145","msg":"Shutting down"}
{"level":"info","ts":1582817948.1367006,"caller":"healthcheck/handler.go:128","msg":"Health Check state change","status":"unavailable"}
{"level":"info","ts":1582817948.136725,"caller":"app/agent.go:89","msg":"shutting down agent's HTTP server","addr":"[::]:5778"}
{"level":"info","ts":1582817948.136871,"caller":"app/agent.go:73","msg":"agent's http server exiting"}
{"level":"info","ts":1582817948.1374054,"caller":"healthcheck/handler.go:128","msg":"Health Check state change","status":"unavailable"}
panic: close of closed channel

goroutine 1 [running]:
github.com/jaegertracing/jaeger/plugin/storage/badger.(*Factory).Close(0xc0001a0180, 0x1df6d01, 0x7f6ba409f120)
	/mnt/storage/jpkroehling/Projects/src/github.com/jaegertracing/jaeger/plugin/storage/badger/factory.go:167 +0x33
github.com/jaegertracing/jaeger/plugin/storage/badger/spanstore.(*SpanWriter).Close(0xc0003b5a40, 0x1df49a0, 0xc0003b5a40)
	/mnt/storage/jpkroehling/Projects/src/github.com/jaegertracing/jaeger/plugin/storage/badger/spanstore/writer.go:197 +0x34
main.main.func1.1()
	/mnt/storage/jpkroehling/Projects/src/github.com/jaegertracing/jaeger/cmd/all-in-one/main.go:156 +0x123
github.com/jaegertracing/jaeger/cmd/flags.(*Service).RunAndThen(0xc00005a980, 0xc0003219a0)
	/mnt/storage/jpkroehling/Projects/src/github.com/jaegertracing/jaeger/cmd/flags/service.go:149 +0x1f6
main.main.func1(0xc000458240, 0x2995420, 0x0, 0x0, 0x0, 0x0)
	/mnt/storage/jpkroehling/Projects/src/github.com/jaegertracing/jaeger/cmd/all-in-one/main.go:150 +0x1627
github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra.(*Command).execute(0xc000458240, 0xc00003c1d0, 0x0, 0x0, 0xc000458240, 0xc00003c1d0)
	/mnt/storage/jpkroehling/Projects/src/github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra/command.go:698 +0x42c
github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc000458240, 0xc000458240, 0xc000215eb8, 0x9)
	/mnt/storage/jpkroehling/Projects/src/github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra/command.go:783 +0x2c9
github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/mnt/storage/jpkroehling/Projects/src/github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra/command.go:736
main.main()
	/mnt/storage/jpkroehling/Projects/src/github.com/jaegertracing/jaeger/cmd/all-in-one/main.go:185 +0x6f7
exit status 2

```

--- a/cmd/collector/app/collector.go
+++ b/cmd/collector/app/collector.go
@@ -16,7 +16,6 @@ package app
 
 import (
 	"context"
-	"io"
 	"net/http"
 	"time"
 
@@ -176,16 +175,6 @@ func (c *Collector) Close() error {
 
 	if err := c.spanProcessor.Close(); err != nil {
 		c.logger.Error("failed to close span processor.", zap.Error(err))
-	}
-
-	// the span processor is closed
-	if c.spanWriter != nil {
-		if closer, ok := c.spanWriter.(io.Closer); ok {
-			err := closer.Close() // SpanWriter
-			if err != nil {
-				c.logger.Error("failed to close span writer", zap.Error(err))
-			}
-		}
 	}
 
 	return nil

--- a/cmd/collector/main.go
+++ b/cmd/collector/main.go
@@ -17,6 +17,7 @@ package main
 
 import (
 	"fmt"
+	"io"
 	"log"
 	"os"
 
@@ -95,6 +96,12 @@ func main() {
 			c.Start(collectorOpts)
 
 			svc.RunAndThen(func() {
+				if closer, ok := spanWriter.(io.Closer); ok {
+					err := closer.Close()
+					if err != nil {
+						logger.Error("failed to close span writer", zap.Error(err))
+					}
+				}
 				c.Close()
 			})
 			return nil

